### PR TITLE
Implement a Time-To-Live (TTL) Indexing feature a la MongoDB

### LINKED ISSUE #346: Support MongoDB TTL feature
It would be awesome if NeDB supported a Time-To-Live (TTL) feature for expiring documents per datastore.

MongoDB has implemented this feature via a special indexing property:
    https://docs.mongodb.org/manual/tutorial/expire-data/

### PR DESCRIPTION
**Pull request by JamesMGreene**
Fixes #346
Closes #347

### CODE REVIEW

**Comment by JamesMGreene** [comment]
cc: @lukasbestle

File: lib/datastore.js [hunk 1]
```diff
@@ -535,9 +579,9 @@ Datastore.prototype._update = function (query, updateQuery, options, cb) {
   }
   , function () {   // Perform the update
     var modifiedDoc
-	  , candidates = self.getCandidates(query)
-	  , modifications = []
-	  ;
+      , candidates = self.getCandidates(query)
```
Comment (JamesMGreene) [inline lib/datastore.js]: Some random tabs that should've been spaces.

File: lib/indexes.js [hunk 2]
```diff
@@ -230,28 +245,21 @@ Index.prototype.revertUpdate = function (oldDoc, newDoc) {
...
 
-// Append all elements in toAppend to array
-function append (array, toAppend) {
-  var i;
-
-  for (i = 0; i < toAppend.length; i += 1) {
-    array.push(toAppend[i]);
-  }
-}
-
```
Comment (JamesMGreene) [inline lib/indexes.js]: Dead code.

File: test/db.test.js [hunk 3]
```diff
@@ -390,7 +390,7 @@ describe('Database', function () {
      *
      * Note: maybe using an in-memory only NeDB would give us an easier solution
      */
-    it('If the callback throws an uncaught execption, dont catch it inside findOne, this is userspace concern', function (done) {
+    it('If the callback throws an uncaught exception, do not catch it inside findOne, this is userspace concern', function (done) {
```
Comment (JamesMGreene) [inline test/db.test.js]: Spelling error in "exception".

File: test/db.test.js [hunk 4]
```diff
@@ -399,11 +399,13 @@ describe('Database', function () {
       process.removeAllListeners('uncaughtException');
 
       process.on('uncaughtException', function MINE (ex) {
+        process.removeListener('uncaughtException', MINE);
```
Comment (JamesMGreene) [inline test/db.test.js]: This listener not being removed was causing me a lot of grief when trying to debug other test failures.  :confounded:

File: test/db.test.js [hunk 5]
```diff
@@ -399,11 +399,13 @@ describe('Database', function () {
...
+
         for (i = 0; i < currentUncaughtExceptionHandlers.length; i += 1) {
           process.on('uncaughtException', currentUncaughtExceptionHandlers[i]);
         }
 
-        ex.should.equal('SOME EXCEPTION');
+        ex.message.should.equal('SOME EXCEPTION');
```
Comment (JamesMGreene) [inline test/db.test.js]: Using proper `Error` instances, a la previous PR #341.

File: test/db.test.js [hunk 6]
```diff
@@ -411,9 +413,9 @@ describe('Database', function () {
         d.findOne({ a : 5}, function (err, doc) {
           if (tryCount === 0) {
             tryCount += 1;
-            throw 'SOME EXCEPTION';
+            throw new Error('SOME EXCEPTION');
```
Comment (JamesMGreene) [inline test/db.test.js]: Using proper `Error` instances, a la previous PR #341.

File: test/db.test.js [hunk 7]
```diff
@@ -411,9 +413,9 @@ describe('Database', function () {
...
           } else {
-            done('Callback was called twice');
+            done(new Error('Callback was called twice'));
```
Comment (JamesMGreene) [inline test/db.test.js]: Using proper `Error` instances, a la previous PR #341.

File: test/indexes.test.js [hunk 8]
```diff
@@ -120,107 +120,126 @@ describe('Indexes', function () {
       assert.deepEqual(idx.tree.search('world'), []);
       assert.deepEqual(idx.tree.search('bloup'), []);
     });
-
```
Comment (JamesMGreene) [inline test/indexes.test.js]: This file diff looks bad but it's really just one new test (lines 123-141), and then properly indenting the "Array fields" suite/description.

File: lib/datastore.js [hunk 9]
```diff
@@ -80,10 +80,37 @@ Datastore.prototype.loadDatabase = function () {
...
 /**
+ * Delete any documents from the array of documents that are expired
+ */
+Datastore.prototype._removeExpiredDocuments = function (docs) {
+  var self = this
+    , expiredDocs = []
+    , validDocs = [];
+
+  docs.forEach(function(doc) {
+    if (model.isExpired(doc) || doc.$$deleted === true) {
```
Comment (JamesMGreene) [inline lib/datastore.js]: @louischatriot: Is needing to check for ` || doc.$$deleted === true` at this point a realistic requirement, or does that field only exist on documents that are written to the disk for later compaction?  I wasn't sure so I erred on the side of caution.

**Comment by JamesMGreene** [comment]
@louischatriot: FWIW, I would love to get this change integrated this week, if possible, as we are shipping a new release next week and this functionality fulfills one of our requirements. :heart:

File: lib/datastore.js [hunk 10]
```diff
@@ -106,14 +133,20 @@ Datastore.prototype.resetIndexes = function (newData) {
...
+  var err
+    , callback = cb || function () {};
 
   options = options || {};
 
-  if (!options.fieldName) { return callback({ missingFieldName: true }); }
+  if (!options.fieldName) {
+    err = new Error("Cannot create an index without a fieldName");
+    err.missingFieldName = true;
+    return callback(err);
```
Comment (JamesMGreene) [inline lib/datastore.js]: Using proper `Error` instances, a la previous PR #341.

File: lib/storage.js [hunk 11]
```diff
@@ -59,11 +59,13 @@ storage.flushToStorage = function (options, callback) {
...
+        var e = null;
         if (errFS || errC) {
-          return callback({ errorOnFsync: errFS, errorOnClose: errC });
-        } else {
-          return callback(null);
+          e = new Error('Failed to flush to storage');
+          e.errorOnFsync = errFS;
+          e.errorOnClose = errC;
         }
+        return callback(e);
```
Comment (JamesMGreene) [inline lib/storage.js]: Using proper `Error` instances, a la previous PR #341.

File: test/db.test.js [hunk 12]
```diff
@@ -89,7 +89,7 @@ describe('Database', function () {
       db = new Datastore({ filename: autoDb, autoload: true, onload: onload })
 
       db.find({}, function (err, docs) {
-        done("Find should not be executed since autoload failed");
+        done(new Error("Find should not be executed since autoload failed"));
```
Comment (JamesMGreene) [inline test/db.test.js]: Using proper `Error` instances, a la previous PR #341.

File: test/executor.test.js [hunk 13]
```diff
@@ -32,8 +33,8 @@ function testThrowInCallback (d, done) {
         done();
       });
     });
-    
-    throw 'Some error';
+
+    throw new Error('Some error');
```
Comment (JamesMGreene) [inline test/executor.test.js]: Using proper `Error` instances, a la previous PR #341.

File: test/executor.test.js [hunk 14]
```diff
@@ -61,13 +62,18 @@ function testRightOrder (d, done) {
...
+                }
+
+                done();
               });
             });
           });
-          
-          throw 'Some error';
+
+          throw new Error('Some error');
```
Comment (JamesMGreene) [inline test/executor.test.js]: Using proper `Error` instances, a la previous PR #341.

File: test/executor.test.js [hunk 15]
```diff
@@ -25,6 +25,7 @@ function testThrowInCallback (d, done) {
   d.find({}, function (err) {    
     process.nextTick(function () {
       d.insert({ bar: 1 }, function (err) {
+        process.removeAllListeners('uncaughtException');
```
Comment (JamesMGreene) [inline test/executor.test.js]: Cleaning up the listener added during the test as well.

File: test/executor.test.js [hunk 16]
```diff
@@ -61,13 +62,18 @@ function testRightOrder (d, done) {
             d.update({ a: 2 }, { a: 3 }, {}, function () {
               d.find({}, function (err, docs) {
                 docs[0].a.should.equal(3);
-                done();
 
+                process.removeAllListeners('uncaughtException');
+                for (var i = 0; i < currentUncaughtExceptionHandlers.length; i += 1) {
+                  process.on('uncaughtException', currentUncaughtExceptionHandlers[i]);
+                }
```
Comment (JamesMGreene) [inline test/executor.test.js]: Cleaning up the listener added during the test and adding back any of the previously attached listeners.

File: test/persistence.test.js [hunk 17]
```diff
@@ -670,7 +670,7 @@ describe('Persistence', function () {
             fs.existsSync(testDb).should.equal(true);
             fs.existsSync(testDb + '~').should.equal(false);
             if (!contents.match(/^{"hello":"world","_id":"[0-9a-zA-Z]{16}"}\n$/)) {
-              throw "Datafile contents not as expected";
+              throw new Error("Datafile contents not as expected");
```
Comment (JamesMGreene) [inline test/persistence.test.js]: Using proper `Error` instances, a la previous PR #341.

File: test/persistence.test.js [hunk 18]
```diff
@@ -697,7 +697,7 @@ describe('Persistence', function () {
...
```
Comment (JamesMGreene) [inline test/persistence.test.js]: Using proper `Error` instances, a la previous PR #341.

File: test/persistence.test.js [hunk 19]
```diff
@@ -721,7 +721,7 @@ describe('Persistence', function () {
...
```
Comment (JamesMGreene) [inline test/persistence.test.js]: Using proper `Error` instances, a la previous PR #341.

File: test/persistence.test.js [hunk 20]
```diff
@@ -743,7 +743,7 @@ describe('Persistence', function () {
         fs.existsSync(dbFile).should.equal(true);
         fs.existsSync(dbFile + '~').should.equal(false);
         if (contents != "") {
-          throw "Datafile contents not as expected";
+          throw new Error("Datafile contents not as expected");
```
Comment (JamesMGreene) [inline test/persistence.test.js]: Using proper `Error` instances, a la previous PR #341.

**Comment by lukasbestle** [comment]
Wouldn't it make sense to put all these housekeeping changes in a separate PR?

Also, the documentation is still about indices and therefore outdated, isn't it?

**Comment by JamesMGreene** [comment]
> Wouldn't it make sense to put all these housekeeping changes in a separate PR?

Yes, the housekeeping grew larger than expected... was hoping to just sneak it in but then I kept finding other little issues. :flushed: 

> Also, the documentation is still about indices and therefore outdated, isn't it?

Nope, that is still how the consumer would invoke this functionality.

File: lib/datastore.js [hunk 21]
```diff
@@ -80,10 +80,37 @@ Datastore.prototype.loadDatabase = function () {
...
+      expiredDocs.push({ $$deleted: true, _id: doc._id });
+      self.removeFromIndexes(doc);
+    }
+    else {
+      validDocs.push(doc);
+    }
+  });
+
+  self.persistence.persistNewState(expiredDocs, function (err) {
+    //return callback(err || null);
```
Comment (JamesMGreene) [inline lib/datastore.js]: Is there anything more appropriate to do here?  Not that I could figure....

File: lib/datastore.js [hunk 22]
```diff
@@ -423,7 +472,7 @@ Datastore.prototype.find = function (query, projection, callback) {
     if (err) { return callback(err); }
 
     for (i = 0; i < docs.length; i += 1) {
-      res.push(model.deepCopy(docs[i]));
+      res.push(model.deepCopy(docs[i], true));
```
Comment (JamesMGreene) [inline lib/datastore.js]: This is necessary to remove the `$$expiry` property before returning documents.

File: lib/datastore.js [hunk 23]
```diff
@@ -459,7 +508,7 @@ Datastore.prototype.findOne = function (query, projection, callback) {
   var cursor = new Cursor(this, query, function(err, docs, callback) {
     if (err) { return callback(err); }
     if (docs.length === 1) {
-      return callback(null, model.deepCopy(docs[0]));
+      return callback(null, model.deepCopy(docs[0], true));
```
Comment (JamesMGreene) [inline lib/datastore.js]: This is necessary to remove the `$$expiry` property before returning documents.

**Comment by JamesMGreene** [comment]
OK, all of the housekeeping changes have been removed from this PR. :gift_heart:

**Comment by JamesMGreene** [comment]
@louischatriot: [As alluded to above](https://github.com/louischatriot/nedb/pull/349#issuecomment-169382995), if there is any chance of this being merged and released before Monday, I'd :yellow_heart: you for it.

**Comment by louischatriot** [comment]
Thanks for the PR, unfortunately very little chance of me merging before Monday. I suggest you use your forked version in the time being and I'll keep you updated.

**Comment by JamesMGreene** [comment]
OK, thanks for the update, @louischatriot.  Will do....

**Comment by JamesMGreene** [comment]
Rebased to current `master`.

File: lib/persistence.js [hunk 24]
```diff
@@ -204,20 +204,24 @@ Persistence.prototype.persistNewState = function (newDocs, cb) {
...
-    var doc;
-
     try {
       doc = model.deserialize(this.beforeDeserialization(data[i]));
       if (doc._id) {
+
+        // Check for expired documents to delete
+        if (doc.$$deleted !== true && model.isExpired(doc)) {
+          doc.$$deleted = true;
+        }
```
Comment (JamesMGreene) [inline lib/persistence.js]: Although this works well, I think I'm actually going to remove it as it prevents me from detecting the removal of expired documents.

Comment (JamesMGreene) [inline lib/persistence.js]: Removed.

File: lib/datastore.js (hunk 9 above)
Comment (JamesMGreene) [inline lib/datastore.js]: Seems to be an unrealistic requirement as a `$$deleted: true` document would've already been removed from all indexes. Removed the second half of the clause from this `if` statement.

File: lib/datastore.js (hunk 21 above)
Comment (JamesMGreene) [inline lib/datastore.js]: Disregard... I now allow an optional callback to be passed in to this function, though I see no immediate need to use it internally within NeDB.  I do have a use case for having it there as a consumer, though, for the sake of extending/overriding NeDB functions to support additional functionality (like emitting events, a la PR #156)... at least for now.

**Comment by louischatriot** [comment]
Hello,

Finally had the time to review, thanks for the PR. Unfortunately I didn't agree with the implementation so I didn't merge but recoded the feature. It is now part of nedb v1.7.1.

**Comment by JamesMGreene** [comment]
****shrugs****

As long as it works, I'm happy to have been the inspiration and catalyst to getting it implemented.  :+1:

**Comment by louischatriot** [comment]
Yeah that's frustrating but at least that's one more great feature in nedb,
one that I actually would have needed for another project not long ago :)
Le 19 janv. 2016 3:17 PM, "James M. Greene" <notifications@github.com> a
écrit :

**Pull request closed without merging**